User guide (#72)
neo4j/neo4j-genai-python: docs/source/user_guide.rst

.. _user-guide:

User Guide
#################

> This guide help you getting started with the Neo4j GenAI Python package,
> and explain how to configure it to meet your requirements.

Reviewer: ```suggestion
This guide will help you get started with the Neo4j GenAI Python package and explain how to configure it to meet your specific needs.
```

************

> Quick start

Reviewer: Very pedantic but should we have Title Case? I do this elsewhere in the documentation. Happy with either as long as we are consistent

Author: You're right, seems to be the convention also for other neo4j docs. I've updated all titles.

    query_text = "How do I do similarity search in Neo4j?"
    response = rag.search(query_text=query_text, retriever_config={"top_k": 5})
    print(response.answer)

Also see :ref:`llminterface`.

Configuring the prompt
========================

Prompt are managed through `PromptTemplate` classes. More
specifically, the `GraphRAG` pipeline uses a `RagTemplate` with
a default prompt. You can use another prompt by subclassing
the `RagTemplate` class and passing it to the `GraphRAG` pipeline
object during initialization:

.. code:: python

    from neo4j_genai.generation import RagTemplate, GraphRAG

    # retriever = ...
    # llm = ...

    prompt_template = RagTemplate(
        prompt="Answer the question {question} using context {context} and examples {examples}",
        expected_inputs=["context", "question", "examples"]
    )

    rag = GraphRAG(retriever=retriever, llm=llm, prompt_template=prompt_template)

    # ...

Also see :ref:`prompttemplate`.

The last piece you can configure in the GraphRAG pipeline is the retriever. The different options
are described below.

************************
Retriever Configuration
************************

> We provide implementation for the most commonly used retrievers:

Reviewer: Could a diagram be useful here in showing the different retrieval strategies?

1. Vector retriever: performs a similarity search based on a Neo4j vector index and a query text or vector. Returns the matched node.
2. Vector Cypher retriever: performs a similarity search based on a Neo4j vector index and a query text or vector. Returned results can be configured through a retrieval query parameter that will be executed after the index search. It can be used to fetch more context around the matched node.
3. External retrievers: use these retrievers when your vectors are not saved in Neo4j but in an external vector database. We currently support Weaviate and Pinecone vector databases.
4. Hybrid and Hybrid cypher retrievers: these retrievers use both a vector and full-text indexes.

> Retrievers all expose a `search` method that we will discuss in the next sections.

Reviewer: Should we comment saying that this is an approximate nearest neighbor search and link to https://neo4j.com/docs/cypher-manual/current/indexes/semantic-indexes/vector-indexes/#_limitiations_and_known_issues ? We also have this in the index page and README

Reviewer: Basically so that developers are aware that this isn't an exact search

Also see :ref:`retrievers`.

Vector retriever
===================

The easiest way to instantiate a vector retriever is:

.. code:: python

    from neo4j_genai.retrievers import VectorRetriever

    retriever = VectorRetriever(
        driver,
        index_name=POSTER_INDEX_NAME,
    )

The `index_name` is the name of the Neo4j vector index that will be used for similarity search.

Search similar vector
-----------------------------

To find the top 3 most similar nodes, you can perform a search by vector:

.. code:: python

    vector = []  # a list of floats, same size as the vectors in the Neo4j vector index
    retriever_result = retriever.search(query_vector=vector, top_k=3)

However, most of the time, you will be given a text (from user) and not a vector. This
use case is covered thanks to the `Embedder`.

Search similar text
-----------------------------

> When searching for a text, you must tell the retriever how to transform (embbed) the text

Reviewer: When searching for a text, you must tell the retriever how to transform (embed) the text
```suggestion
When searching for a text, you must tell the retriever how to transform (embed) the text
```

    retriever = WeaviateNeo4jRetriever(
        driver=driver,
        client=client,
        embedder=embeder,
        collection="Movies",
        id_property_external="neo4j_id",
        id_property_neo4j="id",
    )

Internally, this retriever performs the vector search in Weaviate, finds the corresponding node by matching
the Weaviate metadata `id_property_external` with a Neo4j `node.id_property_neo4j`, and returns the matched node.

Similarly to the vector retriever, you can also use `return_properties` or `retrieval_query` parameters.

Also see :ref:`weaviateneo4jretriever`.

Pinecone retrievers
-------------------

.. note::

    In order to import this retriever, you must install the Weaviate Python client:
    `pip install pinecone-client`

.. code:: python

    from pinecone import Pinecone
    from neo4j_genai.retrievers import PineconeNeo4jRetriever

    client = Pinecone()  # ... create your Pinecone client

    retriever = PineconeNeo4jRetriever(
        driver=driver,
        client=client,
        index_name="Movies",
        id_property_neo4j="id",
        embedder=embeder,
    )

Also see :ref:`pineconeneo4jretriever`.

Other retrievers
===================

> Hybrid and Hybrid Cypher retrievers

Reviewer: Will you add anything to this section? If not perhaps just add what is the difference between vector retrievers i.e. you can search over a fulltext index as well

Author: Added some content, thanks.

- **Hybrid and Hybrid Cypher retrievers**: These retrievers use both a vector and a full-text index in Neo4j.
- **Text to Cypher retriever**: Translate the user question into a Cypher query to be run against a Neo4j database (or Knowledge Graph). The results of the query are then passed to the LLM to generate the final answer.

Retrievers all expose a `search` method that we will discuss in the next sections.

See :ref:`retrievers`.

Vector Retriever
===================

The simplest method to instantiate a vector retriever is:

.. code:: python

    from neo4j_genai.retrievers import VectorRetriever

    retriever = VectorRetriever(
        driver,
        index_name=POSTER_INDEX_NAME,
    )

The `index_name` is the name of the Neo4j vector index that will be used for similarity search.

.. warning::

    Vector index use an **approximate nearest neighbor** algorithm.
    Refer to the `Neo4j Documentation <https://neo4j.com/docs/cypher-manual/current/indexes/semantic-indexes/vector-indexes/#_limitiations_and_known_issues>`_ to learn about its limitations.

> Search Similar Vector

Reviewer: Maybe "Search for Similar Vector" and "Search for Similar Text" below?

    query_text = "How do I do similarity search in Neo4j?"
    retriever_result = retriever.search(query_text=query_text, top_k=3)

Embedders
-----------------------------

Currently, this package supports two embedders: `OpenAIEmbeddings` and `SentenceTransformerEmbeddings`.

The `OpenAIEmbedder` was illustrated previously. Here is how to use the `SentenceTransformerEmbeddings`:

.. code:: python

    from neo4j_genai.embeddings import SentenceTransformerEmbeddings

    embedder = SentenceTransformerEmbeddings(model="all-MiniLM-L6-v2")  # Note: this is the default model

If another embedder is desired, a custom embedder can be created. For example, consider
the following implementation of an embedder that generates random numbers:

.. warning::
    Do not use it in your application :)

.. code:: python

    import random
    from neo4j_genai.embedder import Embedder

    class RandomEmbedder(Embedder):
        def __init__(
            self,
            size: int,
            seed: int = 42,
        ) -> None:
            self.size = size
            random.seed(seed)

        def embed_query(self, text: str) -> list[float]:
            return [
                random.random()

>                 for _ in range(self.size)

Reviewer: Should we provide an actual example rather than one that produces random vectors here? 

I'm quite fine with either way, but wdyt?

Author: I don't have a lot of ideas to be honest, OpenAI + SentenceTransformers are already covering a lot of use cases. But open to any suggestion :-D

Reviewer: ```python
from llama_index.embeddings.ollama import OllamaEmbedding
from neo4j_genai.embedder import Embedder

class OllamaEmbedder(Embedder):
    def __init__(self, ollama_embedding): 
        self.embedder = ollama_embedding
        
    def embed_query(self, text: str) -> list[float]:
        embedding = self.embedder.get_text_embedding_batch(
            [text], show_progress=True
        )
        return embedding[0] 

ollama_embedding = OllamaEmbedding(
            model_name="llama3",
            base_url="http://localhost:11434",
            ollama_additional_kwargs={"mirostat": 0},
        )
embedder = OllamaEmbedder(ollama_embedding)
vector = embedder.embed_query("some text")  
```

Perhaps one with Ollama? This example would work but I'm wondering if it's overcomplicating it

Author: I like it, it gives an example of wrapping LlamaIndex objects as well.

Author: https://github.com/neo4j/neo4j-genai-python/pull/72/commits/ba4d3231ec7820bf526cd011e2e2b97155e747bc

    Ensure that the pre-filtering is stringent enough to prevent query overload.

The currently supported operators are:

- `$eq`: equal.
- `$ne`: not equal.
- `$lt`: less than.
- `$lte`: less than or equal to.
- `$gt`: greater than.
- `$gte`: greater than or equal to.
- `$between`: between.
- `$in`: value is in a given list.
- `$nin`: not in.
- `$like`: LIKE operator case-sensitive.
- `$ilike`: LIKE operator case-insensitive.

Here are examples of valid filter syntaxes and their meaning:

.. list-table:: Filters syntax
   :widths: 80 80
   :header-rows: 1

   * - Filter
     - Meaning
   * - {"year": 1999}
     - year = 1999
   * - {"year": {"$eq": 1999}}
     - year = 1999
   * - {"year": 2000, "title": "The Matrix"}
     - year = 1999 AND title = "The Matrix"
   * - {"$and": [{"year": 2000}, {"title": "The Matrix"}]}
     - year = 1999 AND title = "The Matrix"
   * - {"$or": [{"title": "The Matrix Revolution"}, {"title": "The Matrix"}]}
     - title = "The Matrix" OR title = "The Matrix Revolution"
   * - {"title": {"$like": "The Matrix"}}
     - title CONTAINS "The Matrix"
   * - {"title": {"$ilike": "the matrix"}}
     - toLower(title) CONTAINS "The Matrix"

> See also:ref:`vectorretriever`.

Reviewer: There's an issue with this hyperlink

Author: I thought I had checked all of them... Thanks for double checking, indeed adding the space is enough to fix it.

    retriever = WeaviateNeo4jRetriever(
        driver=driver,
        client=client,
        embedder=embeder,
        collection="Movies",
        id_property_external="neo4j_id",
        id_property_neo4j="id",
    )

Internally, this retriever performs the vector search in Weaviate, finds the corresponding node by matching
the Weaviate metadata `id_property_external` with a Neo4j `node.id_property_neo4j`, and returns the matched node.

The `return_properties` and `retrieval_query` parameters operate similarly to those in other retrievers.

See :ref:`weaviateneo4jretriever`.

Pinecone Retrievers
-------------------

.. note::

    In order to import this retriever, the Pinecone Python client must be installed:
    `pip install pinecone-client`

.. code:: python

    from pinecone import Pinecone
    from neo4j_genai.retrievers import PineconeNeo4jRetriever

    client = Pinecone()  # ... create your Pinecone client

    retriever = PineconeNeo4jRetriever(
        driver=driver,
        client=client,
        index_name="Movies",
        id_property_neo4j="id",
        embedder=embeder,
    )

Also see :ref:`pineconeneo4jretriever`.

Other Retrievers
===================

> Hybrid and Hybrid Cypher Retrievers

Reviewer: It doesn't have to be part of this PR, but we could also expand on these retrievers in a later PR

            metadata={
                "title": record.get('movieTitle'),
                "score": record.get("score"),
            }
        )

    retriever = VectorCypherRetriever(
        driver,
        index_name=INDEX_NAME,
        retrieval_query="MATCH (node)<-[:ACTED_IN]-(p:Person) RETURN node.title as movieTitle, node.plot as movieDescription, collect(p.name) as actors, score",
        result_formatter=result_formatter,
    )

Also see :ref:`vectorcypherretriever`.

.. _vector-databases-section:

Vector Databases
====================

.. note::

    For external retrievers, the filter syntax depends on the provider. Please refer to
    the documentation of the Python client for each provider for details.

.. _weaviate-neo4j-retriever-user-guide:

Weaviate Retrievers
-------------------

.. note::

    In order to import this retriever, the Weaviate Python client must be installed:
    `pip install weaviate-client`

.. code:: python

    from weaviate.connect.helpers import connect_to_local
    from neo4j_genai.retrievers import WeaviateNeo4jRetriever

    client = connect_to_local()
    retriever = WeaviateNeo4jRetriever(
        driver=driver,
        client=client,

>         embedder=embeder,

Reviewer: ```suggestion
        embedder=embedder,
```

        collection="Movies",
        id_property_external="neo4j_id",
        id_property_neo4j="id",
    )

Internally, this retriever performs the vector search in Weaviate, finds the corresponding node by matching
the Weaviate metadata `id_property_external` with a Neo4j `node.id_property_neo4j`, and returns the matched node.

The `return_properties` and `retrieval_query` parameters operate similarly to those in other retrievers.

See :ref:`weaviateneo4jretriever`.

.. _pinecone-neo4j-retriever-user-guide:

Pinecone Retrievers
-------------------

.. note::

    In order to import this retriever, the Pinecone Python client must be installed:
    `pip install pinecone-client`

.. code:: python

    from pinecone import Pinecone
    from neo4j_genai.retrievers import PineconeNeo4jRetriever

    client = Pinecone()  # ... create your Pinecone client

    retriever = PineconeNeo4jRetriever(
        driver=driver,
        client=client,
        index_name="Movies",
        id_property_neo4j="id",

>         embedder=embeder,

Reviewer: ```suggestion
        embedder=embedder,
```
